### DISCUSSION TITLE
Interest in unified condition expressions?

### DISCUSSION THREAD

**Discussion by soc** [flix/flix#5817] [General]
The core idea of [unified condition expressions](https://soc.me/languages/unified-condition-expressions) is to replace the different syntactic forms of

- if expressions,
- pattern matching and pattern guards, and
- other forms like if-lets

with a single construct that scales from simple one-liners to complex pattern matches.

I have implemented it recently in [another language](https://github.com/core-lang/core/commit/37fc7) and I thought it might be a good fit for Flix, too?

Let me know if you are interested in such a contribution!

**Comment by magnus-madsen** [comment]
Give me some time to check it out. How does it compare to

The Ultimate Conditional Syntax - https://cse.hkust.edu.hk/~parreaux/ucs_v1.2.pdf ?

**Reply by soc** [reply]
> Give me some time to check it out.

Sure!

> How does it compare to The Ultimate Conditional Syntax?

Roughly: written 5 years earlier, cites all relevant sources, no interleaved `let`.

**Comment by magnus-madsen** [comment]
I also need to look at https://www.egison.org/

**Reply by soc** [reply]
To clarify: while both unified condition expressions and Egison focus on the same construct, the philosophies and design goals are rather different:

UCE is all about reducing constructs/concepts/complexity while retaining the same expressiveness as before, while Egison spends language complexity to make the language more powerful.

Both are valid choices for a language like Flix to make, but I can only offer expertise on the former, not the latter.

**Comment by magnus-madsen** [comment]
Thanks for the detailed descriptions and write up- I had some more time to think about it. 

Looking at the "language complexity budget" I don't think this is something we want to pursue at the moment. 

On the syntax side, I am more interested in better support for monadic programming, see e.g.:

- https://idris2.readthedocs.io/en/latest/tutorial/interfaces.html#notation
- https://contributors.scala-lang.org/t/quiet-for-comprehensions/6160

I actually like the idea of unified condition expressions, but I just worry about adoption of Flix if too many things are alien and my goal is to sell effect systems (among other things), so that is where I prefer to use the complexity budget. I hope that makes sense.

**Reply by magnus-madsen** [reply]
But this is a discussion thread, so lets keep it up and see if others chime in :)

**Reply by soc** [reply]
Fair enough! (Btw, you don't need to create a new thread for every reply, right? ;-))

**Reply by magnus-madsen** [reply]
Do'h. I guess I am still learning the ropes :)

**Reply by soc** [reply]
(Just one thing I want to clarify because I feel the terms "complexity budget" and "familiarity budget" got mixed up and that's one of my pet peeves; I'll leave you alone after that. :-))

> too many things are alien [...] so that is where I prefer to use the complexity budget

My core design goal with UCE was the net reduction of complexity in that specific language area. The intention is to have *more* complexity budget to spend elsewhere (like on effect systems), not *less*.
 
So the familiarity angle is certainly valid, but I feel it has little intersection with complexity in this case.
